crates/wizer/src/snapshot.rs

```diff
@@ -73,14 +64,12 @@ impl DataSegment {
     ///
     /// `self` must be in front of `other` and they must not overlap with each
     /// other.
-    fn merge(&self, other: &Self) -> DataSegment {
+    fn merge(&mut self, other: &Self) {
         let gap = self.gap(other);
 
-        DataSegment {
-            offset: self.offset,
-            len: self.len + gap + other.len,
-            ..self.clone()
-        }
+        self.len += gap + other.len;
+        self.data.extend(std::iter::repeat(0).take(gap as usize));
+        self.data.extend_from_slice(&other.data);
```

Comment:
Hmm... I'm a little worried about some lurking accidental quadratic stuff in here. I don't have time at this very moment, but if you do, could you look at how `merge` is used below and double check that we don't call merge `O(len(data segments))` times? If all looks good, then feel free to merge.

Otherwise, I can take a closer look on Monday.